mobile/examples/phi-3/ios/LocalLLM/LocalLLM/GenAIGenerator.mm

```diff
@@ -5,45 +5,109 @@
-  const char* prompt = [promptString UTF8String];
-
-  auto sequences = OgaSequences::Create();
-  tokenizer->Encode(prompt, *sequences);
+typedef std::chrono::high_resolution_clock Clock;
+typedef std::chrono::time_point<Clock> TimePoint;
 
-  auto params = OgaGeneratorParams::Create(*model);
-  params->SetSearchOption("max_length", 200);
-  params->SetInputSequences(*sequences);
-
-  // Streaming Output to generate token by token
-  auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);
-
-  auto generator = OgaGenerator::Create(*model, *params);
++ (void)generate:(nonnull NSString*)input_user_question {
+    NSLog(@"Starting token generation...");
+    
+    NSString* llmPath = [[NSBundle mainBundle] resourcePath];
+    const char* modelPath = llmPath.cString;
+    
+    // Log model creation
+    NSLog(@"Creating model ...");
+    auto model = OgaModel::Create(modelPath);
+    if (!model) {
+        NSLog(@"Failed to create model.");
+        return;
+    }
+    
+    NSLog(@"Creating tokenizer...");
+    auto tokenizer = OgaTokenizer::Create(*model);
+    if (!tokenizer) {
+        NSLog(@"Failed to create tokenizer.");
+        return;
+    }
+    
+    auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);
+    
+    // Construct the prompt
+    NSString* promptString = [NSString stringWithFormat:@"<|user|>\n%@<|end|>\n<|assistant|>", input_user_question];
+    const char* prompt = [promptString UTF8String];
+    
+    NSLog(@"Encoding prompt...");
+    auto sequences = OgaSequences::Create();
+    tokenizer->Encode(prompt, *sequences);
+    
+    // Log parameters
+    NSLog(@"Setting generator parameters...");
+    auto params = OgaGeneratorParams::Create(*model);
+    params->SetSearchOption("max_length", 200);
+    params->SetInputSequences(*sequences);
+    
+    NSLog(@"Creating generator...");
+    auto generator = OgaGenerator::Create(*model, *params);
+    
+    bool isFirstToken = true;
+    TimePoint startTime = Clock::now();
+    TimePoint firstTokenTime;
+    int tokenCount = 0;
+    
+    NSLog(@"Starting token generation loop...");
+    while (!generator->IsDone()) {
+        generator->ComputeLogits();
+        generator->GenerateNextToken();
+        
+        if (isFirstToken) {
+            NSLog(@"First token generated.");
+            firstTokenTime = Clock::now();
+            isFirstToken = false;
+        }
+        
+        // Get the sequence data
+        const int32_t* seq = generator->GetSequenceData(0);
+        size_t seq_len = generator->GetSequenceCount(0);
+        
+        // Decode the new token
+        const char* decode_tokens = tokenizer_stream->Decode(seq[seq_len - 1]);
+        
+        // Check for decoding failure
+        if (!decode_tokens) {
+            NSLog(@"Token decoding failed.");
+            break;
+        }
+        
+        NSLog(@"Decoded token: %s", decode_tokens);
```

Comment:
nit: might want a setting to control whether we log inside the loop (if we're after the best perf numbers possible we probably don't want to do that), or a way to exclude time in calls to NSLog from the total.